# <font color="red"> Santander Coders</font>
## <span style="color:red">Lógica de Programação II (Python)</span>

### <span style="color:darkred">Python: coleções aninhadas, comprehension e expressões lambda</span>

*Maurício Luiz Sobrinho*<br>
*Adatech*

---
__Conteúdo:__
- Coleções aninhadas
- Comprehensions
- Expressões lambda

# Coleções aninhadas

Listas, tuplas, dicionários, e outros podem ser aninhados

#### Lista de tuplas

In [1]:
# tuplas como registros indicando frutas: código identificador, nome e preço por kilo
fruta1 = (640, 'morango', 25.0)
fruta2 = (201, 'banana', 4.99)
fruta3 = (452, 'seriguela', 9.99)
fruta4 = (330, 'melancia', 2.50)

lista = [fruta1, fruta2, fruta3, fruta4]
print(lista)

[(640, 'morango', 25.0), (201, 'banana', 4.99), (452, 'seriguela', 9.99), (330, 'melancia', 2.5)]


In [3]:
lista[-1]

(330, 'melancia', 2.5)

Permite usar os benefícios da lista, enquanto mantendo tuplas como registros

In [4]:
lista.append((202, 'maçã', 6.95))
lista.sort()
print(lista)

[(201, 'banana', 4.99), (202, 'maçã', 6.95), (330, 'melancia', 2.5), (452, 'seriguela', 9.99), (640, 'morango', 25.0)]


### Lista de listas

In [7]:
# listas com notas de turmas de alunos
turma1 = [9.5, 8.0, 10.0, 7.0]
turma2 = [1.5, 5.0, 5.5, 6.8, 8.0, 9.5, 9.5, 10.0, 1.0]
turma3 = [5.5, 7.0, 8.0, 6.0, 0.0, 3.5]
turma4 = [10.0, 7.5, 8.0, 9.0, 4.0, 6.0, 6.5]

notas = [turma1, turma2]
print(notas)
notas = notas + [turma3]
print(notas)
notas.append(turma4)
print(notas)

[[9.5, 8.0, 10.0, 7.0], [1.5, 5.0, 5.5, 6.8, 8.0, 9.5, 9.5, 10.0, 1.0]]
[[9.5, 8.0, 10.0, 7.0], [1.5, 5.0, 5.5, 6.8, 8.0, 9.5, 9.5, 10.0, 1.0], [5.5, 7.0, 8.0, 6.0, 0.0, 3.5]]
[[9.5, 8.0, 10.0, 7.0], [1.5, 5.0, 5.5, 6.8, 8.0, 9.5, 9.5, 10.0, 1.0], [5.5, 7.0, 8.0, 6.0, 0.0, 3.5], [10.0, 7.5, 8.0, 9.0, 4.0, 6.0, 6.5]]


#### Aninhando dicionários e listas

In [8]:
notas = {
    'Python': { 2020: [9.5, 8.0, 10.0, 5.0, 6.0, 6.0],
                2021: [8.2, 10.0, 10.0, 9.0, 7.5, 7.0, 10.0] },
    'Redes Neurais': { 2019: [8.0, 9.8, 10.0, 0.0, 8.2, 7.5, 7.5],
                       2020: [9.0, 7.2, 9.0, 8.0, 6.3, 7.0],
                       2021: [10.0, 7.5, 8.0, 9.0, 7.0, 7.3, 6.5] }
}

In [9]:
notas['Python']

{2020: [9.5, 8.0, 10.0, 5.0, 6.0, 6.0],
 2021: [8.2, 10.0, 10.0, 9.0, 7.5, 7.0, 10.0]}

In [16]:
sum(notas['Python'])

4041

In [19]:
type(notas['Python'])

dict

In [20]:
for turma in notas['Redes Neurais']:
    print(turma)

2019
2020
2021


In [21]:
for turma in notas['Redes Neurais'].items():
    print(turma[0], '- Menor nota:', min(turma[1]))

2019 - Menor nota: 0.0
2020 - Menor nota: 6.3
2021 - Menor nota: 6.5


# Comprehension
 
Recurso que permite aplicar uma operação *para cada item de uma sequência* de forma eficiente. 

A sintaxe de um comprehension é:
```
lista = [expressao for variavel_local in objeto]
```
O resultado é equivalente a:
```
lista = []
for variavel_local in objeto:
    lista.append(expressao)
```

Porém, comprehension é executado de forma muito mais rápida. 

Tipicamente se emprega *comprehension* para construir listas ou dicionários.

---

Exemplo: criar uma lista com os valores entre -50 e 50 ao quadrado

In [26]:
%%timeit
quadr = [x**2 for x in range(-50,51)]

20.5 µs ± 279 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)


* `range(-50,51)` gera um intervalo entre -50 e 50
* comprehension eleva ao quadrado cada número `x` nesse intervalo
* o resultado é armazenado na lista `quadr`

In [23]:
print(quadr)

[2500, 2401, 2304, 2209, 2116, 2025, 1936, 1849, 1764, 1681, 1600, 1521, 1444, 1369, 1296, 1225, 1156, 1089, 1024, 961, 900, 841, 784, 729, 676, 625, 576, 529, 484, 441, 400, 361, 324, 289, 256, 225, 196, 169, 144, 121, 100, 81, 64, 49, 36, 25, 16, 9, 4, 1, 0, 1, 4, 9, 16, 25, 36, 49, 64, 81, 100, 121, 144, 169, 196, 225, 256, 289, 324, 361, 400, 441, 484, 529, 576, 625, 676, 729, 784, 841, 900, 961, 1024, 1089, 1156, 1225, 1296, 1369, 1444, 1521, 1600, 1681, 1764, 1849, 1936, 2025, 2116, 2209, 2304, 2401, 2500]


In [27]:
%%timeit
quadr_for = []
for x in range(-50,51):
    quadr_for.append(x**2)

24.2 µs ± 556 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)


In [25]:
print(quadr_for)

[2500, 2401, 2304, 2209, 2116, 2025, 1936, 1849, 1764, 1681, 1600, 1521, 1444, 1369, 1296, 1225, 1156, 1089, 1024, 961, 900, 841, 784, 729, 676, 625, 576, 529, 484, 441, 400, 361, 324, 289, 256, 225, 196, 169, 144, 121, 100, 81, 64, 49, 36, 25, 16, 9, 4, 1, 0, 1, 4, 9, 16, 25, 36, 49, 64, 81, 100, 121, 144, 169, 196, 225, 256, 289, 324, 361, 400, 441, 484, 529, 576, 625, 676, 729, 784, 841, 900, 961, 1024, 1089, 1156, 1225, 1296, 1369, 1444, 1521, 1600, 1681, 1764, 1849, 1936, 2025, 2116, 2209, 2304, 2401, 2500]


Podemos compará-los com o comando `%%timeit`

> É possível omitir a variável local caso essa não seja necessária.

In [30]:
import random as rd

n = 20
rand_num = [rd.randint(1,100) for _ in range(n)]
print(rand_num)

[64, 96, 79, 40, 99, 64, 27, 21, 21, 21, 29, 12, 46, 98, 78, 17, 56, 16, 94, 93]


In [33]:
x = 4.36363
print(round(x,4))

4.3636


---

#### <font color="blue">Exercício 2.1 </font>

Codifique uma função que use comprehension para retornar uma lista com `n` valores numéricos iniciando em 0 e  com passo `p` permitindo um número float como passo. Arredonde cada número para 5 casas decimais usando a função `round(,5)`

Exemplo para n=8, p=0.05
```
[0.0, 0.05, 0.1, 0.15, 0.2, 0.25, 0.3, 0.35, 0.4]
```

---
## Compreehension com filtragem

Combinado com `if` permite filtrar a sequência gerada
```python
variavel = [expressao for variavel_local in objeto if condicao]
```

O código acima é equivalente a:
```
variavel=[]
for variavel_local in objeto:
    if condicao:
        variavel.append(expressao)
```

Gerar o quadrado dos números entre -20 e 20, e montar uma lista apenas com os números ímpares resultantes

In [36]:
l = [x**2 for x in range(-20,21) if (x**2)%2 != 0]
print(l)

[361, 289, 225, 169, 121, 81, 49, 25, 9, 1, 1, 9, 25, 49, 81, 121, 169, 225, 289, 361]


Notar que ali a condição se refere à variável local `x`

#### Iterando por coleções

Vamos criar uma coleção de disciplinas com nome e crédito

In [39]:
disciplinas = [('Programação', 4), ('Cálculo',4), ('Isostática',2), ('Semicondutores', 2),
               ('Manufatura Discreta',2), ('Análise real', 4), ('Seminários', 1), 
               ('Processamento de Imagens', 3)]

In [40]:
mincred = 3

disciplinas_mincred = [(nome,cred) for (nome,cred) in disciplinas if cred >= mincred]
print(disciplinas_mincred)

[('Programação', 4), ('Cálculo', 4), ('Análise real', 4), ('Processamento de Imagens', 3)]


In [41]:
disciplinas_mincred_for = list()
for (nome,cred) in disciplinas:
    if (cred >= mincred):
        disciplinas_mincred_for.append((nome,cred))
print(disciplinas_mincred_for)        

[('Programação', 4), ('Cálculo', 4), ('Análise real', 4), ('Processamento de Imagens', 3)]


In [42]:
disciplinas_mincred == disciplinas_mincred_for

True

### Usando `if-else`

Podemos usar `else` mas nesse caso a estrutura deve vir **antes** do `for`:

```python
variavel = [<expressao> if <condicao> else <expressao_se_falso> for <variavel_local> in <objeto>]
```

> É útil quando temos um valor para substituir no caso em que a condição é falsa

In [46]:
# computando o quadrado dos números entre -20 e 20. Se o resultado for par, substituir por -1
l = [x**2 if (x**2)%2!=0 else -1 for x in range(-20,21)]
print(l)

[-1, 361, -1, 289, -1, 225, -1, 169, -1, 121, -1, 81, -1, 49, -1, 25, -1, 9, -1, 1, -1, 1, -1, 9, -1, 25, -1, 49, -1, 81, -1, 121, -1, 169, -1, 225, -1, 289, -1, 361, -1]


---

**<font color="blue">Exercício 2.2</font>**
A partir de um vetor com números inteiros aleatórios, calcular seu `log` e criar uma lista com o resultado.
* se o número for 0 substituir o valor por `nan` (not a number) da biblioteca `math` para indicar que o resultado não é numérico

In [47]:
import random as rd

n = 25
rand_num = [rd.randint(0,10) for _ in range(n)]
print(rand_num)

[4, 3, 9, 10, 4, 0, 10, 4, 8, 6, 6, 3, 1, 7, 1, 3, 9, 9, 10, 2, 2, 6, 4, 9, 6]


## Comprehensions com iteração paralela em diferentes coleções

Aninhando comprehensions:
```
lista = [<expressao> for <var1, var2, ...> in zip(<colecao1>, <colecao2>, ...)]
```
O resultado do comando acima é equivalente a:
```python
lista = []
for <var1, var2, ...> in zip(<colecao1>, <colecao2>, ...)
    lista.append(expressao)
```

Há várias situações em que é interessante percorrer duas coleções ou sequências

Para isso podemos usar a função `zip()` em que passamos as sequencias para o zip que irá retornar um elemento de cada por vez

```
zip(seq1, seq2, ... seq2n)
```

In [48]:
l1 = list(range(1,10))
l2 = list(range(0,26,2))
t1 = tuple(range(10,100,10))

print(l1)
print(l2)
print(t1)
print()

[1, 2, 3, 4, 5, 6, 7, 8, 9]
[0, 2, 4, 6, 8, 10, 12, 14, 16, 18, 20, 22, 24]
(10, 20, 30, 40, 50, 60, 70, 80, 90)



In [49]:
for i,j,t in zip(l1, l2, t1):
    print(i,j,t)

1 0 10
2 2 20
3 4 30
4 6 40
5 8 50
6 10 60
7 12 70
8 14 80
9 16 90


Note que a  `l2` possui mais elementos, nesse caso o `for` vai nivelar pela sequência de menor tamanho

---
Combinando com comprehensions, vamos usar essa idéia para percorrer duas listas e montar uma nova lista em que copiamos os elementos iguais, e atribuímos `False` nas posições em que as listas são diferentes

In [50]:
vals1 = [5, 5, 5, 1, 2, 3, 4, 6, 7, 7]
vals2 = [5, 5, 5, 1, 2, 4, 4, 6, 6, 7]

equal_pos = [x if x == y else False for (x,y) in zip(vals1, vals2)]
print(equal_pos)

[5, 5, 5, 1, 2, False, 4, 6, False, 7]


---

**<font color="blue">Exercício 2.3</font>**

Codifique uma função que receba como argumento duas listas de números com o mesmo tamanho. Use comprehension para retornar uma nova lista que é a multiplicação elemento-a-elemento das duas listas.  Caso as listas não possuam o mesmo tamanho emita uma mensagem de erro e retorne a constante `math.nan` do módulo `math`.

Exemplo:
```
l1 = [1, 2, 3, 4, 5]
l2 = [5, 5, 5, 10, 10]
multiplica_listas(l1,l2)

  [5, 10, 15, 40, 50]
```

---

## Comprehensions Aninhados (Nested Loops)

Aninhando comprehensions:
```
lista = [<expressao> for <var_local1> in <objeto1> if <condicao1>
                        for <var_local2> in <objeto2> if <condicao2>]
```
O resultado do comando acima é equivalente a:
```python
lista=[]
for var_local1 in objeto1:
    if condicao1:
        for var_local2 in objeto2:
            if condicao2:
                lista.append(expressao)
```

Uma aplicação é gerar o *produto cartesiano* entre duas coleções, ou seja, todas as combinações possíveis das duas coleções.

Mais formalmente, se $A$ e $B$ são conjuntos, o seu produto cartesiano é o conjunto de todos os pares $(a,b)$ em que $a \in A$ e $b \in B$.

In [59]:
A = ['a', 'b']
B = [10, 20, 30]

prod_cart = {(a,b) for a in A for b in B}
print("Com comprehension:", prod_cart)
print(type(prod_cart))

Com comprehension: {('b', 30), ('a', 30), ('b', 20), ('a', 20), ('b', 10), ('a', 10)}
<class 'set'>


In [60]:
# com for
prod_cart_for = set()
for a in A:
    for b in B:
        prod_cart_for.add((a,b))

print("Com for:          ", prod_cart_for)

Com for:           {('b', 30), ('a', 30), ('b', 20), ('a', 20), ('b', 10), ('a', 10)}


#### Formando lista de listas

O elemento do comprehension pode ser uma coleção por exemplo para montar uma lista usamos:

In [62]:
lista1 = [x for x in range(1,6)]
print(lista1)

[1, 2, 3, 4, 5]


Inserindo esse comprehension num outro comprehension podemos repetí-lo

In [69]:
lista_listas = [[x*l for x in range(1,6)] for l in range(1,4)]
print(lista_listas)

[[1, 2, 3, 4, 5], [2, 4, 6, 8, 10], [3, 6, 9, 12, 15]]


Num cenário mais complexo, podemos querer formar uma lista de listas com um padrão, em que cada lista possui valores 0, exceto por um elemento igual a 1 que equivale a sua posição.


Exemplo com 3 listas:
```
[[1, 0, 0],
 [0, 1, 0],
 [0, 0, 1]]
```

Vamos começar produzindo apenas a primeira linha

In [73]:
n = 3
pos = 0
linha = [1 if x == pos else 0 for x in range(n)]
print(linha)

[1, 0, 0]


Agora, como aninhar isso num outro comprehension que permita mudar a a posição?

In [78]:
n = 10
linhas = [[1 if x == pos else 0 for x in range(n)] for pos in range(n)]

for lin in linhas:
    print(lin)


[1, 0, 0, 0, 0, 0, 0, 0, 0, 0]
[0, 1, 0, 0, 0, 0, 0, 0, 0, 0]
[0, 0, 1, 0, 0, 0, 0, 0, 0, 0]
[0, 0, 0, 1, 0, 0, 0, 0, 0, 0]
[0, 0, 0, 0, 1, 0, 0, 0, 0, 0]
[0, 0, 0, 0, 0, 1, 0, 0, 0, 0]
[0, 0, 0, 0, 0, 0, 1, 0, 0, 0]
[0, 0, 0, 0, 0, 0, 0, 1, 0, 0]
[0, 0, 0, 0, 0, 0, 0, 0, 1, 0]
[0, 0, 0, 0, 0, 0, 0, 0, 0, 1]


---

#### <font color="blue">Exercício 2.4 </font>

Codifique uma comprehension que simule uma matriz de tamanho n x n, cujos elementos são dados por `(i+j*i)`, sendo `i` o índice da linha e `j` o índice da coluna. Para simular isso com uma lista de listas, o `i` corresponde ao índice da lista principal e o `j` aos índices das listas aninhadas.

Por exemplo, seja a segunda lista (posição i=1), o seu terceiro elemento (posição j=2) seria obtido por `1+2*1 = 3`

Exemplo com n = 3:
```
  [[0, 0, 0],
   [1, 2, 3],
   [2, 4, 6]] 
```

---

####  <font color="red">Desafio</font>

Temos uma série de pontos 3D organizados numa lista e gostaríamos de computar as distâncias entre todos os pontos pareados.

Calcular a distância entre dois pontos $p1 = (x_1,y_1,z_1)$ e  $p2 = (x_2,y_2,z_2)$ usando a fórmula

$$d(p1,p2) = |x_1 - x_2| +|y_1 - y_2| + |z_1 - z_2|,$$
em que $|.|$ representa o valor absoluto.

Exemplo com 2 pontos organizados em listas:
```
[[1.0, 1.0, 1.0],
 [3.0, 3.0, 3.0]]
```

A saída deve ser:
```
[[0.0, 6.0],
 [6.0, 0.0]]
```

Note que a diagonal principal tem sempre valor zero já que representa a distância de um ponto para ele mesmo, e que a matriz é simétrica pois a distância entre dois pontos p1 e p2 é tal que: d(p1,p2) = d(p2,p1).

Para isso utilize uma única linha contendo comprehensions aninhadas e com iteração em paralelo

In [80]:
from math import fabs, fsum

pontos3d = [ [1.0, 1.0, 1.0],
             [0.0, 1.0, 3.0],
             [2.0, 2.0, 2.0],
             [0.0, 0.0, 0.0] ]

---
# Expressões Lambda
 
Funções **anônimas** que podem ser usadas como *expressão*.


* por ser expressão, não declaração, pode ser usada quando `def` não é possível, por exemplo dentro de um comprehension. 

Sintaxe:
```python
lambda <arg1>, <arg2>, ..., <argn>: <expressao>
```
A `expressao` será o valor de retorno

Um exemplo simples, em que desejamos elevar um número ao quadrado

In [81]:
def quadrado(x):
    return x**2

quadrado(6)

36

Usando uma expressão lambda utilizaríamos

In [82]:
lambda x: x**2

<function __main__.<lambda>(x)>

Mas sendo **anônima** não é possível invocá-la diretamente, precisamos associá-la a um objeto

In [83]:
fquad = lambda x: x**2

fquad(6)

36

**Exemplos** de aplicação incluem seu **uso dentro de comprehensions**

Abaixo um comprehension para avaliar uma lista de números inteiros e criar uma lista contendo 'even' para números pares e 'odd' para números ímpares

In [84]:
# retorna 'even' para x par e 'odd' para x ímpar
f = lambda x: 'even' if (x%2 == 0) else 'odd'

print(3, f(3))
print(10, f(10))

3 odd
10 even


In [85]:
lista_numeros = [1, 1, 2, 2, 10, 11, 12, 13]

odd_or_even = [f(i) for i in lista_numeros]
print(odd_or_even)

['odd', 'odd', 'even', 'even', 'even', 'odd', 'even', 'odd']


In [86]:
filter_odd = [i for i in lista_numeros if f(i) == 'odd']
print(filter_odd)

[1, 1, 11, 13]


---
### Expressão lambda como parte de argumento para função

Em algumas funções podemos passar por argumento uma expressão.

Por exemplo ao ordenar elementos numa lista

In [87]:
cientistas_da_computacao = ['Betty Holberton', 'Alan Turing', 'Dennis Ritchie', 'Grace Hopper']

_ = [print(i) for i in cientistas_da_computacao]

Betty Holberton
Alan Turing
Dennis Ritchie
Grace Hopper


In [88]:
cientistas_da_computacao.sort()
print(cientistas_da_computacao)

['Alan Turing', 'Betty Holberton', 'Dennis Ritchie', 'Grace Hopper']


Gostaríamos de ordenar pelo sobrenome!

O método `sort()` possui um argumento `key` para definir o critério de ordenação


In [89]:
help(cientistas_da_computacao.sort)

Help on built-in function sort:

sort(*, key=None, reverse=False) method of builtins.list instance
    Sort the list in ascending order and return None.
    
    The sort is in-place (i.e. the list itself is modified) and stable (i.e. the
    order of two equal elements is maintained).
    
    If a key function is given, apply it once to each list item and sort them,
    ascending or descending, according to their function values.
    
    The reverse flag can be set to sort in descending order.



In [90]:
# usando como chave o tamanho
cientistas_da_computacao.sort(key=len)
print(cientistas_da_computacao)

['Alan Turing', 'Grace Hopper', 'Dennis Ritchie', 'Betty Holberton']


Ou com uma **lambda expression** que retorne o último nome

In [93]:
print(cientistas_da_computacao[0].split())

['Alan', 'Turing']


In [95]:
print(cientistas_da_computacao[2].split()[-1].lower())

ritchie


In [96]:
cientistas_da_computacao.sort(key = lambda nome: nome.split()[-1].lower())
print(cientistas_da_computacao)

['Betty Holberton', 'Grace Hopper', 'Dennis Ritchie', 'Alan Turing']


> Isso seria impossível usando `def`

---

#### <font color="blue">Exercício 2.5</font>

Escreva uma expressão lambda que permita receber uma string contendo duas palavras relativas ao nome de uma pessoa e seu sobrenome. Essas são separadas por um ou mais espaços em branco. A expressão deve retornar o nome no formato: "SOBRENOME, N."

Exemplo:
```
nome = 'Dennis   ritchie'
```

Retorno:
```
RITCHIE, D.
```

A seguir escreva um comprehension que percorra uma lista contendo nomes e gere uma nova lista com os nomes no formato "SOBRENOME, N."

Exemplo:
```
nomes = ['Dennis   ritchie', 'ALAN  Turing', 'betty Holberton']
```

Retorno:
```
['RITCHIE, D.', 'TURING, A.', 'HOLBERTON, B.']
```
